### PR TITLE
Fix reduced_index for Integer UnitRanges having eltypes other than Int

### LINKED ISSUE #39995: Bug in evaluating reduced indices with BigInt UnitRange arguments
```julia
julia> reduce(+, big(1):big(2), dims = 1)
ERROR: TypeError: in typeassert, expected Tuple{Base.OneTo{BigInt}}, got a value of type Tuple{Base.OneTo{Int64}}
Stacktrace:
 [1] reduced_indices(inds::Tuple{Base.OneTo{BigInt}}, d::Int64)
   @ Base ./reducedim.jl:23
 [2] reduced_indices(a::UnitRange{BigInt}, region::Int64)
   @ Base ./reducedim.jl:15
 [3] reducedim_initarray(A::UnitRange{BigInt}, region::Int64, init::BigInt, #unused#::Type{BigInt})
   @ Base ./reducedim.jl:92
 [4] _reducedim_init(f::typeof(identity), op::typeof(+), fv::typeof(zero), fop::typeof(sum), A::UnitRange{BigInt}, region::Int64)
   @ Base ./reducedim.jl:124
 [5] reducedim_init(f::Function, op::typeof(+), A::UnitRange{BigInt}, region::Int64)
   @ Base ./reducedim.jl:109
 [6] _mapreduce_dim(f::Function, op::Function, #unused#::Base._InitialValue, A::UnitRange{BigInt}, dims::Int64)
   @ Base ./reducedim.jl:336
 [7] #mapreduce#710
   @ ./reducedim.jl:322 [inlined]
 [8] #reduce#712
   @ ./reducedim.jl:371 [inlined]
 [9] top-level scope
   @ REPL[3]:1
```

This appears to be due to https://github.com/JuliaLang/julia/blob/c9c1d3e56c958d8d6985261705731c733bed6bdd/base/reducedim.jl#L6 which hardcodes an `Int` value irrespective of the `eltype` of the `OneTo`, which conflicts with 

```julia
julia> axes(big(1):big(2), 1) |> typeof
Base.OneTo{BigInt}
```

### PR DESCRIPTION
**Pull request by jishnub**
Fixes #39995 

Now
```julia
julia> reduce(+, big(1):big(2), dims = 1)
1-element Vector{BigInt}:
 3
```

### CODE REVIEW

**Review by timholy — APPROVED** [review]

**Comment by jishnub** [comment]
This also fixes some other reductions that use non-`Int` axes

Before (using the 32-bit build of Julia):
```julia
julia> Int
Int32

julia> reduce(+, Int64(5):Int64(6), dims = 1)
ERROR: TypeError: in typeassert, expected Tuple{Base.OneTo{Int64}}, got a value of type Tuple{Base.OneTo{Int32}}
```

After:
```julia
julia> reduce(+, Int64(5):Int64(6), dims = 1)
1-element Array{Int64,1}:
 11
```

**Comment by kshyatt** [comment]
Is this good to go?

**Comment by jishnub** [comment]
Yes, this is ready

**Review by johnnychen94 — APPROVED** [review]

**Comment by jishnub** [comment]
Gentle bump, would be nice to have this fixed

File: base/reducedim.jl [hunk 1]
```diff
@@ -3,7 +3,7 @@
 ## Functions to compute the reduced shape
 
 # for reductions that expand 0 dims to 1
-reduced_index(i::OneTo) = OneTo(1)
+reduced_index(i::OneTo) = typeof(i)(1)
```
Comment (simeonschaub) [inline base/reducedim.jl:6]: Just seems a little bit easier to read to me this way:
```suggestion
reduced_index(i::OneTo{T}) where {T} = OneTo(one(T))
```

**Review by simeonschaub — APPROVED** [review]

**Pull request merged**